### PR TITLE
Refactor and cleanup the RegExp parser

### PR DESCRIPTION
**Pull request by szledan**
JerryScript-DCO-1.0-Signed-off-by: Szilard Ledan szledan.u-szeged@partner.samsung.com

### CODE REVIEW

File: jerry-core/parser/regexp/re-parser.cpp [hunk 1]
```diff
@@ -27,150 +27,157 @@
...
- * @return unicode codeunit
+ * @return true, if callback function holds true for lookup number of character ahead
+ *         false, otherwise
  */
-static ecma_char_t
+static bool
 re_lookup (lit_utf8_iterator_t iter, /**< input string iterator */
-           uint32_t lookup) /**< size of lookup */
+           uint32_t lookup, /**< size of lookup */
+           bool (*char_is_digit_function) (ecma_char_t)) /**< callback function */
```
Comment (zherczeg) [inline jerry-core/parser/regexp/re-parser.cpp]: Am I understand correctly, that only hex digit is used for char_is_digit_function? I think we don't need specialization here then. Just a re_hex_peek() function

File: jerry-core/parser/regexp/re-parser.cpp [hunk 2]
```diff
@@ -27,150 +27,157 @@
...
 
-  return ch;
+  return is_digit;
 } /* re_lookup */
 
 /**
+ * Consume non greedy (question mark) character if present.
+ *
+ * @return false, if non greedy character found
+ *         true, otherwise
```
Comment (zherczeg) [inline jerry-core/parser/regexp/re-parser.cpp]: I think this function returns false is greedy character is found.

File: jerry-core/parser/regexp/re-parser.cpp [hunk 3]
```diff
@@ -27,28 +27,48 @@
 /**
  * Lookup a character in the input string.
  *
- * @return unicode codeunit
+ * @return true, if lookup number of characters ahead are hax digits
```
Comment (kkristof) [inline jerry-core/parser/regexp/re-parser.cpp]: `hax` typo

File: jerry-core/parser/regexp/re-parser.cpp [hunk 4]
```diff
@@ -562,183 +543,164 @@ re_parse_char_class (re_parser_ctx_t *parser_ctx_p, /**< number of classes */
...
-               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 2))
-               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 3)))
+      else if (ch == LIT_CHAR_LOWERCASE_X
+               && re_hex_lookup (*iter_p, 2))
       {
-        lit_utf8_iterator_advance (&(parser_ctx_p->iter), 2);
-
         lit_code_point_t code_point;
-        const lit_utf8_byte_t *hex_start = parser_ctx_p->iter.buf_p + parser_ctx_p->iter.buf_pos.offset;
+        const lit_utf8_byte_t *hex_start = iter_p->buf_p + iter_p->buf_pos.offset;
```
Comment (kkristof) [inline jerry-core/parser/regexp/re-parser.cpp]: Please use `hex_start_p`.

File: jerry-core/parser/regexp/re-parser.cpp [hunk 5]
```diff
@@ -562,183 +543,164 @@ re_parse_char_class (re_parser_ctx_t *parser_ctx_p, /**< number of classes */
...
-               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 4))
-               && lit_char_is_hex_digit (re_lookup (parser_ctx_p->iter, 5)))
+      else if (ch == LIT_CHAR_LOWERCASE_U
+               && re_hex_lookup (*iter_p, 4))
       {
-        lit_utf8_iterator_advance (&(parser_ctx_p->iter), 2);
-
         lit_code_point_t code_point;
-        const lit_utf8_byte_t *hex_start = parser_ctx_p->iter.buf_p + parser_ctx_p->iter.buf_pos.offset;
+        const lit_utf8_byte_t *hex_start = iter_p->buf_p + iter_p->buf_pos.offset;
```
Comment (kkristof) [inline jerry-core/parser/regexp/re-parser.cpp]: Same as above.

**Comment by szledan** [comment]
I have updated the PR.

**Comment by egavrin** [comment]
Please fix precommit.

**Comment by szledan** [comment]
Fixed, sorry.

**Comment by zherczeg** [comment]
LGTM

**Comment by egavrin** [comment]
`make push`

**Pull request merged**